File: day15_part2_2016.dart

import 'dart:io';

void main() {
  List<List<int>> discs = [];
  List<int> positions = [];
  List<int> initialPositions = [];
  List<int> initialDiscs = [];
  int time = 0;

  File file = new File('input.txt');
  List<String> lines = file.readAsLinesSync();

  for (String line in lines) {
    List<String> parts = line.split(' ');
    int discNumber = int.parse(parts[1].substring(1));
    int positionsCount = int.parse(parts[3]);
    int initialPosition = int.parse(parts[parts.length - 1].substring(0, parts[parts.length - 1].length - 1));

    initialDiscs.add(discNumber);
    initialPositions.add(positionsCount);
    positions.add(positionsCount);
    discs.add([discNumber, positionsCount, initialPosition]);
  }

  while (true) {
    bool success = true;
    for (int i = 0; i < discs.length; i++) {
      if ((discs[i][2] + time + i + 1) % discs[i][1] != 0) {
        success = false;
        break;
      }
    }

    if (success) {
      break;
    }

    time++;
  }

  print(time);

  discs.add([initialDiscs.length + 1, 11, 0]);
  positions.add(11);
  initialDiscs.add(initialDiscs.length + 1);
  initialPositions.add(11);

  time = 0;

  while (true) {
    bool success = true;
    for (int i = 0; i < discs.length; i++) {
      if ((discs[i][2] + time + i + 1) % discs[i][1] != 0) {
        success = false;
        break;
      }
    }

    if (success) {
      break;
    }

    time++;
  }

  print(time);
}

File: training_data/reflection-tuning/day15_part2_2016.py
def parse_input(filename):
    discs = []
    with open(filename, 'r') as file:
        for line in file:
            parts = line.split()
            positions = int(parts[3])
            start_pos = int(parts[-1][:-1])
            discs.append((positions, start_pos))
    return discs

def capsule_passes(discs, start_time):
    for i, (positions, start_pos) in enumerate(discs):
        time = start_time + i + 1
        if (start_pos + time) % positions != 0:
            return False
    return True

def find_solution(discs):
    time = 0
    while True:
        if capsule_passes(discs, time):
            return time
        time += 1

# Part One
discs = parse_input('input.txt')
part1_solution = find_solution(discs)
print(f"Part One: {part1_solution}")

# Part Two
discs.append((11, 0))
part2_solution = find_solution(discs)
print(f"Part Two: {part2_solution}")
